### PR TITLE
[Variants] Added support for components with variant templates

### PR DESCRIPTION
**Pull request by juanmanuelramallo**
- Solves #34 
- Updated base object to care about the variant in the request
- Added unit and integration tests to cover this

### CODE REVIEW

**Comment by juanmanuelramallo** [comment]
@joelhawksley I would appreciate your input on this. 

It doesn't look that much clean to me, so let me know if I can improve anything or whether it's worth the inclusion of this into the prod code, because it definitely adds some complexity to the codebase.

**Comment by joelhawksley** [comment]
@juanmanuelramallo thanks for the PR! This is definitely directionally correct.

I'm wondering if we could have `compile` just compile _all_ of the variants at once, instead of one-by-one. Perhaps that would be simpler.

I'd be more than happy to pair on this if you'd like, email me any time: joelhawksley@github.com

**Comment by juanmanuelramallo** [comment]
> @juanmanuelramallo thanks for the PR! This is definitely directionally correct.
> 
> I'm wondering if we could have `compile` just compile _all_ of the variants at once, instead of one-by-one. Perhaps that would be simpler.
> 
> I'd be more than happy to pair on this if you'd like, email me any time: [joelhawksley@github.com](mailto:joelhawksley@github.com)

Yep, we can compile it all at once, but **where would you save the compiled templates? In a class attribute hash maybe?**

Currently, the compiled templates live as the last statement of the `call` method. As I wasn't sure about how the `ActionView::OutputBuffer.new` works, I didn't want to mess it up. That's why I decided to use the `call_#{variant}` methods approach.

I'll try to read about that buffer and check if I can make the same `call` method work for all variants.

Thanks for you answer @joelhawksley!

File: lib/action_view/component/base.rb [hunk 1]
```diff
@@ -102,17 +103,28 @@ def source_location
...
         # Right now this just compiles the template the first time the component is rendered.
         def compile
           return if @compiled && ActionView::Base.cache_template_loading
           ensure_initializer_defined
+          set_compiled_templates
 
-          class_eval <<-RUBY, __FILE__, __LINE__ + 1
-            def call
+          class_eval <<-'RUBY', __FILE__, __LINE__ + 1
+            def call(variant = nil)
```
[the remark this answers is not in the document]
Comment (juanmanuelramallo) [reply lib/action_view/component/base.rb]: So you mean like this? https://github.com/github/actionview-component/pull/86/commits/9ed942d420ac4a034da4232a65e98b3f4951095a#diff-4ea3d395eff742fa0d8952d23a16d377R122

File: test/action_view/component_test.rb [hunk 2]
```diff
@@ -25,14 +25,32 @@ def test_raises_error_when_sidecar_template_is_missing
...
 
     assert_includes error.message, "More than one template found for TooManySidecarFilesComponent."
   end
 
+  def test_raises_error_when_more_than_one_sidecar_template_for_a_variant_is_present
+    error = assert_raises StandardError do
+      render_inline(TooManySidecarFilesForVariantComponent)
+    end
+
+    assert_includes error.message, "More than one template found for a variant in TooManySidecarFilesForVariantComponent."
```
[the remark this answers is not in the document]
Comment (juanmanuelramallo) [reply test/action_view/component_test.rb]: Yep, it's included in the message I think but I didn't include it in here. 
I'll update it

File: test/app/controllers/application_controller.rb [hunk 3]
```diff
@@ -1,4 +1,8 @@
 # frozen_string_literal: true
 
 class ApplicationController < ActionController::Base
+  def variants
+    request.variant = :tablet if params[:tablet]
```
[the remark this answers is not in the document]
Comment (juanmanuelramallo) [reply test/app/controllers/application_controller.rb]: Sounds good

File: test/app/views/application/variants.html.erb [hunk 4]
```diff
@@ -0,0 +1 @@
+<span><%= render VariantsComponent %></span>
```
[the remark this answers is not in the document]
Comment (juanmanuelramallo) [reply test/app/views/application/variants.html.erb]: No need for span you're right

File: lib/action_view/component/base.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (joelhawksley) [reply lib/action_view/component/base.rb]: Yes! Sorry, I somehow managed to submit my review hours after I meant to. My apologies!

[the remark this answers is not in the document]
Comment (juanmanuelramallo) [reply lib/action_view/component/base.rb]: Haha no prob! Thanks for the review

**Comment by joelhawksley** [comment]
Closing in favor of https://github.com/github/actionview-component/pull/92, which builds on this branch. 

Thank you for the work here @juanmanuelramallo ❤️

**Pull request closed without merging**